In [14]:
# -*- coding: utf-8 -*-
from codecs import open
import sys

import sklearn
import pandas as pd
import numpy as np
import seaborn as sns 
from scipy import stats
import matplotlib
matplotlib.use("module://mplcairo.qt")
import matplotlib.pyplot as plt
from matplotlib.font_manager import FontProperties
from IPython.display import set_matplotlib_formats
set_matplotlib_formats('retina', quality=100)
def f1(precision,recall):
    return (2.0*precision*recall)/(precision+recall)

def main(path_goldstandard, path_outputfile):

    truth_dict={}
    output_dict_correct={}
    output_dict_attempted={}
    predicted = {}
    correct = 0
    truth_file_lines=open(path_goldstandard,encoding='utf8').readlines()
    submission_file_lines=open(path_outputfile,encoding='utf8').readlines()
    if len(submission_file_lines)!=len(truth_file_lines): sys.exit('ERROR: Number of lines in gold and output files differ')
    or_distribution = [0 for _ in range(20)]
    pr_distribution = [0 for _ in range(20)]
    for i in range(20):
        predicted[i] = [0 for _ in range(20)]
    for i in range(len(submission_file_lines)):
        line = submission_file_lines[i]
        emoji_code_gold = truth_file_lines[i]
        emoji_code_output=int(submission_file_lines[i].replace("\n",""))
        emoji_code_gold=int(truth_file_lines[i].replace("\n",""))
        if emoji_code_gold == emoji_code_output:
            correct += 1
        or_distribution[emoji_code_gold] += 1
        pr_distribution[emoji_code_output] += 1
        
        predicted[emoji_code_gold][emoji_code_output] += 1
    
    for i in range(len(predicted)):
        total = sum(predicted[i])
        for j in range(len(predicted)):
            predicted[i][j] = str(round(predicted[i][j]/total*100,3)) + "%"
    
    return predicted, or_distribution, pr_distribution, correct


In [15]:
path_goldstandard = "gold.labels"
path_outputfile = "predictions.txt"
predicted_distribution, or_distribution, pr_distribution, correct = main(path_goldstandard, path_outputfile)
print("accuracy: " + str(correct/sum(or_distribution)))
print(predicted_distribution)
print(sum(or_distribution))

accuracy: 0.29932
{0: ['83.932%', '2.436%', '6.566%', '0.315%', '1.584%', '0.232%', '0.241%', '0.482%', '0.0%', '0.009%', '1.445%', '0.333%', '1.009%', '0.0%', '0.065%', '0.157%', '0.0%', '1.176%', '0.019%', '0.0%'], 1: ['75.445%', '5.859%', '10.745%', '0.062%', '2.319%', '0.166%', '0.331%', '0.911%', '0.0%', '0.0%', '1.925%', '0.269%', '1.014%', '0.0%', '0.062%', '0.021%', '0.0%', '0.87%', '0.0%', '0.0%'], 2: ['36.745%', '2.206%', '52.272%', '0.0%', '3.088%', '0.199%', '0.706%', '0.265%', '0.0%', '0.0%', '2.272%', '0.463%', '0.772%', '0.0%', '0.11%', '0.176%', '0.0%', '0.728%', '0.0%', '0.0%'], 3: ['85.758%', '2.303%', '5.489%', '0.422%', '1.42%', '0.115%', '0.192%', '0.806%', '0.0%', '0.0%', '1.536%', '0.154%', '0.883%', '0.0%', '0.038%', '0.115%', '0.0%', '0.768%', '0.0%', '0.0%'], 4: ['36.868%', '2.637%', '19.833%', '0.0%', '31.405%', '0.323%', '0.565%', '1.346%', '0.0%', '0.027%', '4.171%', '0.538%', '1.103%', '0.0%', '0.108%', '0.646%', '0.0%', '0.323%', '0.108%', '0.0%'], 5: ['6

In [16]:
mapping_lines=open("us_mapping.txt",encoding='utf8').readlines()
mapping  = {}
emoji = []
emoji_text = []
for line in mapping_lines:
    temp = line.strip("\n").strip(" ").strip("_").split("\t")
    mapping[int(temp[0])] = [temp[1], temp[2]]
print(mapping)

for i in range(len(mapping)):
    emoji.append(mapping[i][0])
    emoji_text.append(mapping[i][1])
print(emoji)

{0: ['❤', '_red_heart_'], 1: ['😍', '_smiling_face_with_hearteyes_'], 2: ['😂', '_face_with_tears_of_joy_'], 3: ['💕', '_two_hearts_'], 4: ['🔥', '_fire_'], 5: ['😊', '_smiling_face_with_smiling_eyes_'], 6: ['😎', '_smiling_face_with_sunglasses_'], 7: ['✨', '_sparkles_'], 8: ['💙', '_blue_heart_'], 9: ['😘', '_face_blowing_a_kiss_'], 10: ['📷', '_camera_'], 11: ['🇺🇸', '_United_States_'], 12: ['☀', '_sun_'], 13: ['💜', '_purple_heart_'], 14: ['😉', '_winking_face_'], 15: ['💯', '_hundred_points_'], 16: ['😁', '_beaming_face_with_smiling_eyes_'], 17: ['🎄', '_Christmas_tree_'], 18: ['📸', '_camera_with_flash_'], 19: ['😜', '_winking_face_with_tongue_']}
['❤', '😍', '😂', '💕', '🔥', '😊', '😎', '✨', '💙', '😘', '📷', '🇺🇸', '☀', '💜', '😉', '💯', '😁', '🎄', '📸', '😜']


In [17]:


font_path = matplotlib.font_manager.findSystemFonts(fontpaths=None, fontext='ttf')
path = ""
for i in font_path:
    if "seguiemj" in i:
        path = i

myfont = FontProperties(fname=path,size=15)
flist = [f.name for f in matplotlib.font_manager.fontManager.ttflist]
font = matplotlib.font_manager.FontProperties(family='Segoe UI Emoji')
file = matplotlib.font_manager.findfont(font)

In [32]:

predicted_list =[]
for i in range(20):
    predicted_list.append(predicted_distribution[i])
'''Function for displaying histogram of variable.'''
def table(variable, labels):
    

    
    plt.rcParams['figure.figsize'] = (8, 5)
    plt.rc('font', family='Segoe UI Emoji')
    fig, ax = plt.subplots()
    fig.patch.set_visible(False)
    ax.axis('off')
    ax.axis('tight')
    the_table = ax.table(cellText=variable,
                      rowLabels=labels,
                      colLabels=labels,
                      loc='center',
                    cellLoc='center')
    
    prop = FontProperties(fname=file)
    #ax.set_xticklabels(emoji, fontproperties=prop, fontsize = 30)
    # Make the chart fill out the figure better.
    the_table.auto_set_font_size(False)
    the_table.set_fontsize(15)
    the_table.scale(1, 1) 
    
    #plt.tight_layout()
    #plt.savefig(savename, bbox_inches='tight')
    plt.show()
    
    return the_table
table(predicted_list,emoji)

[['431.5', '1049.4', '799.6', '2149.8', '917.9'], ['292.2', '717.8', '456.4', '1368.5', '865.6'], ['213.8', '636.0', '305.7', '1175.2', '796.0'], ['124.6', '555.4', '153.2', '677.2', '192.5'], ['66.4', '174.3', '75.1', '577.9', '32.0']]
{0: ['83.932%', '2.436%', '6.566%', '0.315%', '1.584%', '0.232%', '0.241%', '0.482%', '0.0%', '0.009%', '1.445%', '0.333%', '1.009%', '0.0%', '0.065%', '0.157%', '0.0%', '1.176%', '0.019%', '0.0%'], 1: ['75.445%', '5.859%', '10.745%', '0.062%', '2.319%', '0.166%', '0.331%', '0.911%', '0.0%', '0.0%', '1.925%', '0.269%', '1.014%', '0.0%', '0.062%', '0.021%', '0.0%', '0.87%', '0.0%', '0.0%'], 2: ['36.745%', '2.206%', '52.272%', '0.0%', '3.088%', '0.199%', '0.706%', '0.265%', '0.0%', '0.0%', '2.272%', '0.463%', '0.772%', '0.0%', '0.11%', '0.176%', '0.0%', '0.728%', '0.0%', '0.0%'], 3: ['85.758%', '2.303%', '5.489%', '0.422%', '1.42%', '0.115%', '0.192%', '0.806%', '0.0%', '0.0%', '1.536%', '0.154%', '0.883%', '0.0%', '0.038%', '0.115%', '0.0%', '0.768%', '0

In [21]:
'''Function for displaying histogram of variable.'''
def histo(variable, x_name, y_name, title, labels, savename):
    plt.rcParams['figure.figsize'] = (8, 5)
    plt.rc('font', family='Segoe UI Emoji')
    fig, ax = plt.subplots()

    bar = ax.bar(x=np.arange(len(variable)), 
            height=variable,
            tick_label=emoji
    )
    prop = FontProperties(fname=file)
    ax.set_xticklabels(emoji, fontproperties=prop, fontsize = 30)
    # Make the chart fill out the figure better.
    
    
    plt.xlabel('%s' %x_name, fontsize=18)
    plt.ylabel('%s' %y_name, fontsize=18)
    plt.title('%s' %title, fontsize=35)
    
    plt.grid(True)
    for rect1, label in zip(bar, labels):
        height = rect1.get_height()
        plt.annotate(
            str(rect1.get_height()),
            (rect1.get_x() + rect1.get_width()/2, height+5),
            ha="center",
            va="bottom",
            fontsize=15,
            fontproperties=prop
        )
    
    plt.tight_layout()
    #plt.savefig(savename, bbox_inches='tight')
    plt.show()
    
    return bar


# Actual Label Distribution in the testset 

In [22]:
bar = histo(or_distribution, "emoji", "# of tweets", "Actual Emoji Distribution", emoji, "emojis.png")

<img src="emojis.png">
There are totally 20 labels and 50k tweets. The Label distribution shows that the top 5 highest number of emojis are red_heart❤ 10798, smiling_face_with_hearteyes 😍 4830, face_with_tears_of_joy😂 4534, fire🔥 3716, sparkles✨ 2749. The least used 3 emojis are face_blowing_a_kiss😘 1175, purple_heart💜 1114, winking_face_with_tongue😜 1010.

# Prediction distribution

In [30]:
bar = histo(pr_distribution, "emoji", "# of tweets","Predicted Emoji Distribution", emoji, "pemojis.png")

<img src="pemojis.png">

In [23]:
def Pie(labels, sizes, index):
    
    prop = FontProperties(fname=file)
    colors = ['gold', 'yellowgreen', 'lightcoral', 'lightskyblue']
    explode = [0 for _ in range(20)]  # explode 1st slice
    explode[index] = 1
    sorted_index = sorted(range(len(sizes)), key=sizes.__getitem__)[len(sizes)-5:len(sizes)]
    print(sorted_index)
    for i in sorted_index:
        explode[i] = 0.5
    explode = tuple(explode)
    print(explode)
    # Plot
    patches, labelss, autopct = plt.pie(sizes, explode=explode, labels=labels, autopct='%1.2f%%', shadow=True,   textprops={'fontsize': 14, 'fontproperties':prop})
    for lab in labelss:
        lab.set_fontsize(10)
    plt.legend(patches, labels, loc="best", prop=prop, fontsize = 30)
    plt.axis('equal')
    plt.show()

In [78]:
path_text = "us_test.text"
path_labels = "us_test.labels"
path_predict = "predictions.txt"
list_text = []
list_labels = []
label_to_text = {}
predict_to_text = {}
text_lines=open(path_text,encoding='utf8').readlines()
label_lines=open(path_labels,encoding='utf8').readlines()
predict_lines=open(path_predict,encoding='utf8').readlines()
for i in range(len(label_lines)):
    emoji_code=int(label_lines[i].replace("\n",""))
    emoji_predict = int(predict_lines[i].replace("\n",""))
    text = text_lines[i].strip("\n")
    if emoji_code not in label_to_text:
        label_to_text[emoji_code] = [text]
    else:
        label_to_text[emoji_code].append(text)
    if emoji_predict not in predict_to_text:
        predict_to_text[emoji_predict] = [text]
    else:
        predict_to_text[emoji_predict].append(text)



In [51]:
import math
def Donut(labels, sizes, index):
    title = "Predicted Distribution over tweets with acutal emoji " + labels[i]
    fig, ax = plt.subplots(figsize=(8, 8), subplot_kw=dict(aspect="equal"))
    explode = [0 for _ in range(20)]  # explode 1st slice
    explode[index] = 0.3
    explode = tuple(explode)
    prop = FontProperties(fname=file)
    patches, labelss, autopct = ax.pie(sizes, labels=labels, wedgeprops=dict(width=0.4), autopct='%1.2f%%', startangle=40, explode=explode, 
                           textprops={'fontsize': 30, 'fontproperties':prop})
    sorted_index = sorted(range(len(sizes)), key=sizes.__getitem__)[len(sizes)-10:len(sizes)]
    #print(sorted_index)
    
    for lab, value in zip(labelss,autopct):
        #print(lab)
        
        label_index = labels.index(lab.get_text())
        percent = float(value.get_text().strip("%"))
        if (label_index in sorted_index and percent >1) or label_index == index:
            lab.set_fontsize(15)
        else:
            lab.set_text(" ")
            value.set_text(" ")
    plt.title('%s' %title, fontsize=15)
    plt.tight_layout()
    plt.show()
    plt.savefig("Figure_%d.png"%(index+1))
    plt.close()
    

In [32]:
i = 0
Donut(emoji, predicted_distribution[i], i)

<img src="Figure_1.png">
<img src="Figure_2.png">
<img src="Figure_3.png">
<img src="Figure_4.png">
<img src="Figure_5.png">
<img src="Figure_6.png">
<img src="Figure_7.png">
<img src="Figure_8.png">
<img src="Figure_9.png">
<img src="Figure_10.png">
<img src="Figure_11.png">
<img src="Figure_12.png">
<img src="Figure_13.png">
<img src="Figure_14.png">
<img src="Figure_15.png">
<img src="Figure_16.png">
<img src="Figure_17.png">
<img src="Figure_18.png">
<img src="Figure_19.png">
<img src="Figure_20.png">

# Analysis

## Obeservation:
❤ have 83.932% predicted correctly  
😍 have 5.859% predicted correctly  
😂 have 52.272% predicted correctly  
💕 have 0.422% predicted correctly  
🔥 have 31.405% predicted correctly  
😊 have 1.798% predicted correctly  
😎 have 1.854% predicted correctly  
✨ have 5.966% predicted correctly  
💙 have 0.0% predicted correctly  
😘 have 0.085% predicted correctly  
📷 have 41.76% predicted correctly  
🇺🇸 have 8.415% predicted correctly  
☀ have 21.028% predicted correctly  
💜 have 0.0% predicted correctly  
😉 have 0.613% predicted correctly  
💯 have 3.457% predicted correctly  
😁 have 0.0% predicted correctly  
🎄 have 48.867% predicted correctly  
📸 have 0.29% predicted correctly  
😜 have 0.0% predicted correctly  

For each tweets no matter what emoji at least 30% probability to predict ❤.  

For red heart, blue_heart, purple_heart, double_heart, face_blowing_a_kiss, United_Status which are ❤,💕,💙,💜,😘,🇺🇸 mainly predict ❤ (at least 80 percent, while 🇺🇸 have 8 percent predict right)

For each tweets 😂 is the second emoji that the model tends to predict unconditionally. Especially the model tends to always predict tweets with actual emoji 😜,😁,😉,😘 to 😂 except the effect of ❤

😍 have 5.859% predicted correctly which is second highest accuracy among face emojis. 

For tweets have emoji 📸, the model always predict 📷

📷,🎄,🔥,☀ have some higher accuracy (more than 20%), while ✨,🇺🇸,💯 have relatively low accuracy

## Reasoning:

The model cares too much about the frequency of emojis.

There are no too much context difference between emojis like 💕,💙,💜,😘 with ❤  

There are no too much context difference between emojis like 😜,😁,😉,😘 with 😂, maybe because they all have meaning like joy, happy, laughin

There are no context difference between 📸 and 📷. People only use the two emoji interchangeablely. Nearly no information can show the difference.

📷,🎄,🔥,☀ have its special meaning which is shown in the context with some keywords maybe, while ✨,🇺🇸,💯 don't have strong keywords to show the difference.

## Examples support reasoning:


In [68]:
print(emoji)
print(emoji_text)

['❤', '😍', '😂', '💕', '🔥', '😊', '😎', '✨', '💙', '😘', '📷', '🇺🇸', '☀', '💜', '😉', '💯', '😁', '🎄', '📸', '😜']
['_red_heart_', '_smiling_face_with_hearteyes_', '_face_with_tears_of_joy_', '_two_hearts_', '_fire_', '_smiling_face_with_smiling_eyes_', '_smiling_face_with_sunglasses_', '_sparkles_', '_blue_heart_', '_face_blowing_a_kiss_', '_camera_', '_United_States_', '_sun_', '_purple_heart_', '_winking_face_', '_hundred_points_', '_beaming_face_with_smiling_eyes_', '_Christmas_tree_', '_camera_with_flash_', '_winking_face_with_tongue_']


In [77]:
for i in predict_to_text[17][:10]:
    print(i)

Merry Christmas from Burt Marketing Group. @ Roseburg, Oregon
Happy Thanksgiving, ya turkeys... :: jalen.hutchinson @ Thanksgiving's Heroes
That Christmas concert was L I T @ Lake Highlands High
Drive through the #lights #tgif #weekend #christmastime #friday #lightshow @ Winnebago Park,…
Rehearsal clip. Downbeat at 7:30. Come. Merry Christmas. ️ #ericadicegliemusic @ St. Jerome…
Life is good! The holiday cheer is here #nycjunkgirl #holidayparty #holidays #christmas…
White Christmas @ Washington Ave - Memorial, Houston
Time to place your bakery Christmas orders We’ll be making Yule Logs (chocolate cake with…
A fun day finding our perfect Christmas tree! Athena approves #christmastree #december…
All I want for Christmas I got @ Annapolis, Maryland


In [ ]:
for i in predict_to_text[17][:10]:
    print(i)

In [79]:
for i in predict_to_text[7][:10]:
    print(i)

Tomorrow I'll be at #TheBeatAuction junxioncomplex Check the flyer for more info. @ The…
~ The world needs more sparkle @user #NYFW ~ @ Skylight Clarkson Sq
post show munchies thebeehive_la @ Infinite Energy Center
New clients Expires 1/31/18#socialmediamarketing #couponcommunity #hairstyles #hairgoddess…
Pinking of you Be the first to check out our newest cordless lamps #BSGGlowMini at our booth…
Don’t let society tell you how to be. Express yourself proudly @ Hollywood
Last minute #IVYLOVE POP-UP TODAY at theskinroom104 in #whittier! See @user for details …
A successful night #paradeoflights @ 9NEWS Parade of Lights
We're coming up on 2018, the year of #11 if 2017 wasn't everything you expected, 2018 is your…
Surround yourself with those who make you smile, build you up, and embrace the sunshine @user


In [80]:
for i in predict_to_text[11][:10]:
    print(i)

Polls are closing soon please continue vote for me ️ ️ ️ @ West…
One short month until we can change the location back to the dirty T @ Naperville, Illinois
-Stay tuned!--#startup #2018 #goals #mindset #millionaire #millionairemindest…
Overwatch eSports League coming January’s 10th 2018 •So far there are a total of 12 teams…
Look like Africa #look #africa #la #california @ Los Angeles,…
What could be more American? #4thofjuly #merica #aafnation #america #harleydavidson…
Do it. Wear it. Work it. Appreciate it. Happy Election Day! @user
We are LIVE at the 6th Annual Latino Festival in Timonium, MD!!! #MagentaPrime #BeMagenta …
Voting is cool yo #democracy #vote @ Middleburg, Virginia
Party like it's 1776 #mericamonday @ Buhlow Lake Recreation Area


In [81]:
for i in predict_to_text[15][:10]:
    print(i)

what goes up must come down. #dcmetrolife (pc to @user ) @ Washington Metro
My boy Phene been holding it down for the #717 #PA @ Harrisburg, Pennsylvania
Be yourself, everyone else is taken. #confidence #beyourself @ Lake Travis
Let's get your Thrive on! #heartbeat #healthyfood #health #painrelief #mentalclarity…
Knockin' TF out once I touch down @ Frederick Historic District
facts @ South Carolina
Hustle #SWEATLIFE #weightlifting #fitfam #lifestyle #bodybuilding #lifting #gains #athlete…
Facts be you- nobody built like you never let someone change you…
In order to get better you have to WORK no matter what level your playing on #ffba #hardwork…
falco_tp one of our sound engineers making sure the levels are good #tbt by…
